### ISSUE TITLE
I guess mamba.step could be deleted if selective_scan_fn can accept ssm_state as an input param.

### ISSUE THREAD

**Issue by agiwave** [state-spaces/mamba#233]
Maybe there are some benifit below:
1, The code could be simplier.
2, The inference could be faster.
3, The inference can accept multi-tokens in this way.
There are some reference codes here.  https://github.com/agiwave/Mamba/blob/main/Mamba.py

**Comment by tridao** [comment]
Yep, in some sense `step` is the specialized version of `forward` that accepts ssm_state and only move by 1 step.

**Comment by agiwave** [comment]
Mamba is really a great work on model size control. On this point, there's another thinking. The mamba vocab embedding size is 50280*1024(370m) = about 50m parameters. But it really need d_model=1024 to encode a word? My assumption is: The vocab-dim can be reduce to 128(or even 64) to encode a word. The latent_dim(d_model) can still be 1024 or more to contain more info only a word. So, could have in_proj and out_proj to mapping between vocab_dim and d_model. Maybe it also can give us:
1, Sharing vocab_embedding between diff-size models: 370m, 790m .......
2, Vocab_embbeding size will be cut from  50280*1024 to 50280*128, or maybe 250000*128 will be better.
I can't play LLM for resource limit reason:). So, this can only be an assumption.

**Comment by agiwave** [comment]
Maybe that version above can help a lot on this point. I don't know whether our owner   will take it into consideration. So, i tried it.

**Comment by agiwave** [comment]
Maybe that version above can help a lot on this point. I don't know whether our owner will take it into consideration. So, i tried it.

**Comment by AnaRhisT94** [comment]
Hi @agiwave ,
How can the inference accept multi-tokens in the new updated code of Mamba?
Here is the function of the step:
```
def step(self, hidden_states, conv_state, ssm_state):
        dtype = hidden_states.dtype
        #assert hidden_states.shape[1] == 1, "Only support decoding with 1 token at a time for now"
        xz = self.in_proj(hidden_states.squeeze(1))  # (B 2D)
        x, z = xz.chunk(2, dim=-1)  # (B D)

        # Conv step
        if causal_conv1d_update is None:
            conv_state.copy_(torch.roll(conv_state, shifts=-1, dims=-1))  # Update state (B D W)
            conv_state[:, :, -1] = x
            x = torch.sum(conv_state * rearrange(self.conv1d.weight, "d 1 w -> d w"), dim=-1)  # (B D)
            if self.conv1d.bias is not None:
                x = x + self.conv1d.bias
            x = self.act(x).to(dtype=dtype)
        else:
            x = causal_conv1d_update(
                x,
                conv_state,
                rearrange(self.conv1d.weight, "d 1 w -> d w"),
                self.conv1d.bias,
                self.activation,
            )

        x_db = self.x_proj(x)  # (B dt_rank+2*d_state)
        dt, B, C = torch.split(x_db, [self.dt_rank, self.d_state, self.d_state], dim=-1)
        # Don't add dt_bias here
        dt = F.linear(dt, self.dt_proj.weight)  # (B d_inner)
        A = -torch.exp(self.A_log.float())  # (d_inner, d_state)

        # SSM step
        if selective_state_update is None:
            # Discretize A and B
            dt = F.softplus(dt + self.dt_proj.bias.to(dtype=dt.dtype))
            dA = torch.exp(torch.einsum("bd,dn->bdn", dt, A))
            dB = torch.einsum("bd,bn->bdn", dt, B)
            ssm_state.copy_(ssm_state * dA + rearrange(x, "b d -> b d 1") * dB)
            y = torch.einsum("bdn,bn->bd", ssm_state.to(dtype), C)
            y = y + self.D.to(dtype) * x
            y = y * self.act(z)  # (B D)
        else:
            y = selective_state_update(
                ssm_state, x, dt, A, B, C, self.D, z=z, dt_bias=self.dt_proj.bias, dt_softplus=True
            )

        out = self.out_proj(y)
        return out.unsqueeze(1), conv_state, ssm_state
```

**Comment by AnaRhisT94** [comment]
Ok I made it work finally
